Add multi window code and shader editors (GSOC'22 Project)

trollodel opened the pull request:
This is the main PR for my GSOC'22 project ([link to the project description](https://summerofcode.withgoogle.com/programs/2022/projects/4cIGjw9I)).
Old commit history with expanded commits: [https://github.com/trollodel/godot/tree/gsoc_2022_multiwindow_full_history](https://github.com/trollodel/godot/tree/gsoc_2022_multiwindow_full_history)

The project consists in the following major features:
- Script Editor undocking.
- Shader Editor undocking.
- Window layout saves.

You can read more about it in the GSOC'22 progress report: https://godotengine.org/article/gsoc-2022-progress-report-1

The PR is feature complete, further enhancement will be done in a separate PR, after the merge of this one, which will happen after the 4.0 release as marked by the milestone.

<details>
 <summary>Screenshots/Videos</summary>
 
 https://user-images.githubusercontent.com/33117082/175552167-1ab95e3a-0955-465a-afbf-bc021b53f3d4.mp4

 https://user-images.githubusercontent.com/33117082/179208386-6db8ad98-4351-4a45-9959-66863337b741.mp4

 https://user-images.githubusercontent.com/33117082/179216725-ae5c1fd6-1d0b-477e-8836-c0966fe6afc6.mp4

 https://user-images.githubusercontent.com/33117082/179759817-94b10f29-2f29-4ce1-a458-a7e3a92599fd.mp4 
 
 The screen selector in the Script Editor (two 1920x1080 screen)
![](https://user-images.githubusercontent.com/33117082/185341416-45bf5c96-3bfb-4689-b92c-5a1dea3e91d5.png)

The screen selector with different screen sizes (one 1920x1080 screen and one 1280x960 screen, 4:3 aspect ratio)
![](https://user-images.githubusercontent.com/33117082/185342196-517273ea-c1eb-4841-851a-f4b6cddbbf08.png)

Everything undocked!
![](https://user-images.githubusercontent.com/33117082/185355385-083ca96a-22d5-4c29-b476-658e3d418a7c.png)

</details>

Part of godotengine/godot-proposals#28.

Happy testing!

Supersedes #62963.

*Keywords for easy searching: detach script, detachable, detached, pop out, popping out, other window, separate window, undock, undocked, undockable, floating*

jcostello commented:
Looking good so far. As I mention in the other PR discussion: F1 is not working inside the code window while its focus

jcostello commented:
> @jcostello sorry for the late response, but now global shortcuts (like F1 or Ctrl+S) should work in the script window too.

Nice. Im out til the weekend. I will test it then :)

trollodel commented:
The PR is now feature complete for what I have planned. The only thing remaining to do are a bunch of editor settings to customize the behavior.
Some commits done here will be extracted in dedicated PRs as they are independent, but required, changes.
To indicate the progresses, I'll mark the PR as ready for review, even if it's not ready to be merged (starting from the verbose commit history).

YeldhamDev commented:
Could you please rebase this to the latest `master` commit if possible?

YeldhamDev commented:
Icons in the "Search Help" dialog appear to be broken:
![image](https://user-images.githubusercontent.com/30739239/185659385-6cb3f52b-2198-4224-8db5-a0c39697059e.png)

trollodel commented:
@YeldhamDev fixed. There are probably other places where themes are not properly handled.

YeldhamDev commented:
Here's some (kinda late) feedback:
- Add borders to both the script and shader windows (while the script UI already has borders, they are small compared to the borders of other windows).
![image](https://user-images.githubusercontent.com/30739239/185767014-4e59237b-f0f7-46da-8ff5-3eac91b5e766.png)
- It would be ideal if the shader button on the bottom panel could not be toggled. I don't know how hacky would the solution be, however.
![image](https://user-images.githubusercontent.com/30739239/185767081-70ff2b13-eab5-4dbc-bba2-edc31ecb3c67.png)

YeldhamDev commented:
- There are some [code styling issues](https://github.com/godotengine/godot/runs/7948811356?check_suite_focus=true) that need fixing.
- Thinking again, the extra margins can be disabled for the script editor.

trollodel commented:
@YeldhamDev sorry for the late response

> Add borders to both the script and shader windows (while the script UI already has borders, they are small compared to the borders of other windows).
>
> Thinking again, the extra margins can be disabled for the script editor.

For the shader editor and the docks, margins were added (TBH, docks already has margins).
I created a dedicated stylebox for the script editor to use when it's undocked, to have the same margin as the shader editor.

> It would be ideal if the shader button on the bottom panel could not be toggled. I don't know how hacky would the solution be, however.

Yeah, I give a try, and it would be very hacky to do in the Shader editor. The easiest hack would be to tell EditorNode to not show the bottom panel. The other ugly solution would be disabling the button, but it's not obvious that the button is disabled because the panel is in a window.

YeldhamDev commented:
> For the shader editor and the docks, margins were added (TBH, docks already has margins).

That's the opposite of what I asked. :stuck_out_tongue: I said that the script editor doesn't need the borders anymore, just the shader editor. The borders for the docks are also unnecessary.

Also, rebase to the latest `master` commit, this will resolve a problem with the CI.

trollodel commented:
> That's the opposite of what I asked. I said that the script editor doesn't need the borders anymore, just the shader editor. The borders for the docks are also unnecessary.

I don't agree, I think that all the undockables needs some margins, or they looks bad (and if all the margin can be the same, even better).

Here are the screenshots for comparison:

<details>
 <summary>Script Editor</summary>
 
 1. No margins, no internal margins.
![script_nomargin](https://user-images.githubusercontent.com/33117082/186870735-c293bb36-69bc-4e32-8e16-63f380b167f5.png)

2. Only default margins (the ones used when it's not undocked)
![script_default_margin](https://user-images.githubusercontent.com/33117082/186870898-158af9a7-bbb1-4ce1-b222-393acbbe888d.png)

3. Current behavior, no internal margins.
![script_current](https://user-images.githubusercontent.com/33117082/186871334-225944fa-5bca-4c57-a7c9-afa982854357.png)
</details>

<details>
 <summary>Shader Editor</summary>
 
 4. No margins
![shader_nomargin](https://user-images.githubusercontent.com/33117082/186871504-53c260fa-e46a-4fa5-a126-5faed7a78b8c.png)

5. Current behavior
![shader_current](https://user-images.githubusercontent.com/33117082/186871556-bcace69f-38cb-4d1c-95ca-caad4c75cc37.png)

</details>

<details>
 <summary>Docks</summary>
 
 6. No margins
![dock_nomargin](https://user-images.githubusercontent.com/33117082/186871656-89a6a2ad-41f0-4297-8bfd-fa318f004dc3.png)

7. Current behavior (same as master)
![dock_current_master](https://user-images.githubusercontent.com/33117082/186871672-2407255e-6ecf-4898-b66b-234770354186.png)
</details>

YeldhamDev commented:
> I don't agree, I think that all the undockables needs some margins, or they looks bad (and if all the margin can be the same, even better).

I didn't mean removing the dock borders completely, just that they don't need extra ones.

But from what I see in the screenshots, everything looks fine with the current changes.

trollodel committed:
Add multi window code and shader editors

clayjohn commented:
I tested this locally and it appears to work fine. I was able to pop out the script and shader editors, move them around, and close them and nothing broke. 

My only nitpick is the placement of the "Make floating" button on the shader editor. In the shader editor it is tucked into the right-hand side, while on the script editor it is placed on the left. My preference is for it to be on the left for both. But, this is a very weak preference, I am more than happy to go along with the crowd on this one. 

![Screenshot from 2022-11-09 09-35-30](https://user-images.githubusercontent.com/16521339/200901245-289b5190-b982-4f03-bd14-8744629c741b.png)
![Screenshot from 2022-11-09 09-35-26](https://user-images.githubusercontent.com/16521339/200901249-3c7b7ec8-5d09-4db3-a7ba-d0b66a020e90.png)

I read through the code as well watching out for issues that I might recognize from the rendering side of things and nothing jumped out at me. It appears this is mostly contained within editor code (with which I am pretty unfamiliar)

trollodel commented:
@clayjohn thanks for the testing.

> My only nitpick is the placement of the "Make floating" button on the shader editor. In the shader editor it is tucked into the right-hand side, while on the script editor it is placed on the left. My preference is for it to be on the left for both. But, this is a very weak preference, I am more than happy to go along with the crowd on this one.

I'm not happy with the placement of that button in both the Script and Shader editor, but I failed to find a better place for them. 
Moving it to the left along other menus would make it out of place. It's more consistent on the right.

Kemeros commented:
How about putting it on the right hand side and hide it behind a triple vertical dots menu?

![Screenshot_20221109-131648_Firefox Beta](https://user-images.githubusercontent.com/15658976/200909502-8208a48d-cfa1-4c2c-a300-24c8ae556d28.jpg)

Geometror reviewed (commented):
Tested this and it works overall pretty well. Great work! 
There is just one thing which does not work currently: Shortcuts (all floating windows are affected) 
Seems like they are handled by the main window, which is a regression of this PR since the floating docks which are already available (tested with beta 5) accept the shortcuts correctly when focused.

Regarding the "Make floating" button: I'm also in favor of placing it on the right, however I have no opinion on whether it should be hidden behind a ellipsis menu. Doing so would create some room for more options that can be added later on, but currently we don't need it. Maybe we could make the icon a bit more clear (although it's already pretty unambiguous) and remove the text (add it as a tooltip)? That makes it a bit more compact.

akien-mga commented:
> Maybe we could make the icon a bit more clear (although it's already pretty unambiguous) and remove the text (add it as a tooltip)? That makes it a bit more compact.

I agree that this could be icon only, like the distraction free/full screen buttons. We should strive to keep the central toolbars compat enough so that the editor can be used even on screens with low resolution and/or in windowed mode.

trollodel commented:
@Geometror thanks for testing.

> There is just one thing which does not work currently: Shortcuts (all floating windows are affected)
Seems like they are handled by the main window, which is a regression of this PR since the floating docks which are already available (tested with beta 5) accept the shortcuts correctly when focused.

Can you give me an example of a broken shortcut?

> Regarding the "Make floating" button: I'm also in favor of placing it on the right, however I have no opinion on whether it should be hidden behind a ellipsis menu. Doing so would create some room for more options that can be added later on, but currently we don't need it. Maybe we could make the icon a bit more clear (although it's already pretty unambiguous) and remove the text (add it as a tooltip)? That makes it a bit more compact.

I'll make some mockups to help choose.

Geometror commented:
You can pick any shortcut for the floating dock/editor, none of them work. For example, Ctrl+A to select all text, Ctrl+Z/Ctrl+Shift+Z for Undo/Redo, etc.

trollodel commented:
For all people interested, here are the mockups:

<details>
	<summary>Mockups</summary>

	Mixed alignments, icon with text (current)
   ![mixed_text](https://user-images.githubusercontent.com/33117082/202759422-0ff4afbe-bb1c-4d96-9ba0-86fd50bad64b.png)
	Mixed alignments, icon only (ignore the dock's button, should be icon only)
	![mixed_icon](https://user-images.githubusercontent.com/33117082/202759655-89083b86-0c33-4e1f-aedd-1c49dbfcff41.png)
	Right alignments, icon with text
	![right_text](https://user-images.githubusercontent.com/33117082/202760064-88cb956d-fd8c-49eb-b048-d9503be162f6.png)
	Right alignments, icon only (ignore the dock's button, should be icon only)
	![right_icon](https://user-images.githubusercontent.com/33117082/202760187-65170378-7fa0-4341-a4c4-e651c3f6bb45.png)
</details>

To me, icon only it's inconsistent with the other button, especially in the script editor.

> Maybe we could make the icon a bit more clear (although it's already pretty unambiguous) 

Suggestions welcome. I'm not good at doing icon/2d draw.

trollodel commented:
> You can pick any shortcut for the contained dock/editor in a floating window, none of them work. For example, Ctrl+A to select all text, Ctrl+Z/Ctrl+Shift+Z for Undo/Redo, etc.

I can reproduce it.

Geometror commented:
Can confirm that it works nicely now :)

YuriSizov commented:
> There is just one thing which does not work currently: Shortcuts (all floating windows are affected)
Seems like they are handled by the main window, which is a regression of this PR since the floating docks which are already available (tested with beta 5) accept the shortcuts correctly when focused.

I would ask for a bit more investigation into this issue, because it is definitely not a regression from this PR. Existing windows also fail to capture shortcuts, for example I've seen this happen with the manage dialog of the Theme editor. It is still reproduceable in beta 5.

So unless this PR fixes this for every editor window, we need to look into fixing this globally. If a certain setup is required for each window node to work, then it needs to be documented.

trollodel commented:
> > There is just one thing which does not work currently: Shortcuts (all floating windows are affected)
> > Seems like they are handled by the main window, which is a regression of this PR since the floating docks which are already available (tested with beta 5) accept the shortcuts correctly when focused.
> 
> I would ask for a bit more investigation into this issue, because it is definitely not a regression from this PR. Existing windows also fail to capture shortcuts, for example I've seen this happen with the manage dialog of the Theme editor. It is still reproduceable in beta 5.

The issue described by Geometror is different. The issue was that every shortcut was routed to the main window.
IIUC, what you describe is that input events (including shortcuts) emitted in any other window are not propagated to the main window. I think it's expected, because propagating input events to the main window by default is undesired.

> So unless this PR fixes this for every editor window, we need to look into fixing this globally. If a certain setup is required for each window node to work, then it needs to be documented. It is possible that it's the window in the Theme editor is missing something specifically, but it seems to be an easy trap for developers in that case.

What is done in this PR is to add a special node (added as first child of the window, as described [here](https://docs.godotengine.org/en/latest/tutorials/inputs/inputevent.html#how-does-it-work)) to capture all the unhandled shortcut inputs and route them to the main window. You probably need to do the same to the theme editor.

YuriSizov commented:
> What is done in this PR is to add a [special node](https://github.com/godotengine/godot/pull/62378/files#diff-238093108f35fbbedb6402d8824b903da3d8323167b2ee83b08baafee9b80d44R41-R73) (added as first child of the window, as described [here](https://docs.godotengine.org/en/latest/tutorials/inputs/inputevent.html#how-does-it-work)) to capture all the unhandled shortcut inputs and route them to the main window. You probably need to do the same to the theme editor.

Thanks, I'll look into it. That's a lot of boilerplate, though. I wonder if we can streamline it. From my quick testing, the project settings dialog doesn't block shortcuts. Would be nice to not have to replicate this every time.

trollodel commented:
> From my quick testing, the project settings dialog doesn't block shortcuts.

Strange. To me, it doesn't work on both the PS dialog and the theme dialog (tested with F1). 
To clarify, shortcuts that are handled locally in the window (like Ctrl+F for search fields) are not affected by the issue I described above. If they are broken, then it's another bug.

YuriSizov commented:
> Strange. To me, it doesn't work on both the PS dialog and the theme dialog (tested with F1).
> To clarify, shortcuts that are handled locally in the window

I've tested the same thing, undoing changes made in the theme editor. Unless I got turned around and confused, that at least worked with the project settings open, but didn't work with the theme editor's manage dialog (which is kind of ironic, give the other way around would be less bad).

Well, we can look into it separately, your fix for this PR is definitely a useful data point.

trollodel commented:
Feel free to ping me in RC if you want to discuss it. The PM dialog is doing sus stuff with UndoRedo.

akien-mga commented:
I'll try to get this reviewed and tested further so we can hopefully merge it in the near future.
Seems like it will need another rebase to solve some merge conflicts.

KoBeWi commented on editor/editor_help_search.cpp:
```diff
@@ -117,8 +117,12 @@ void EditorHelpSearch::_notification(int p_what) {
+		} break;
+
+		case NOTIFICATION_THEME_CHANGED:
+		case NOTIFICATION_ENTER_TREE: {
```
`NOTIFICATION_THEME_CHANGED` is now received automatically on enter tree, so this is no longer necessary.

akien-mga replied:
Are you sure about this? I remembered there was a PR ensuring this, but IIRC it introduces regressions and was reverted.

KoBeWi replied:
https://github.com/godotengine/godot/blob/e0e93ce094d19d132f34aa5c92e892524a8125c9/scene/main/window.cpp#L1155

trollodel replied:
Can confirm that works without that line as suggested.

KoBeWi replied:
Then you can revert this whole file, no? The connection can be done in enter tree as before.

trollodel replied:
No, because it may enter tree multiple times and `connect` print an error when the signal is already connected.

KoBeWi replied:
This should be `NOTIFICATION_THEME_CHANGED` if the only thing it does is updating icons.
What I meant with the first comment is that this notification is received when entering tree. But it can also be received afterwards when the theme changes, so ENTER_TREE is not sufficient here.

trollodel replied:
I can do that, but do note that this PR only moves the `connect` call into NOTIFICATION_READY.

KoBeWi replied:
I know, but now the theme changed notification can be used here.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -4652,67 +4653,71 @@ void EditorNode::_copy_warning(const String &p_str) {
-	window->set_meta("dock_index", dock_index);
-	gui_base->add_child(window);
+	Size2 dock_size = p_dock->get_size() + borders * 2;
+	Point2 dock_screen_pos = p_dock->get_global_position() + get_tree()->get_root()->get_position() - borders;
```
```suggestion
	Point2 dock_screen_pos = p_dock->get_screen_position() - borders;
```
I think it's the same.

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,398 @@
+
+	window = memnew(Window);
+	window->set_wrap_controls(true);
+	window->set_transient(false);
```
This is already false by default.

trollodel replied:
I put it just to be sure, but I removed it.

trollodel commented:
> I did some testing and it works fine, but I found one (minor?) issue. When you close and re-open a window, it will briefly appear at the old location and then gets forced to the "starting location" (i.e. its position in the main window). I think if the window appears at a remembered position, it could as well stay there.

I test it again and the feature is completely broken now, and put the window in the top left of the screen.

The intended feature is to place the window to match the control rect when it's docked. I didn't think about remembering the last window rect.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -4652,67 +4653,71 @@ void EditorNode::_copy_warning(const String &p_str) {
+
+	wrapper->set_window_enabled(true);
+
+	wrapper->connect("window_visibility_changed", callable_mp(this, &EditorNode::_dock_floating_close_request).bind(wrapper));
```
You can use `close_requested` signal for this.

trollodel replied:
WindowWrapper doesn't have the `close_requested` signal. Maybe I can add it just for this situation.

KoBeWi replied:
Its Window has this signal. You can either make window publicly accessible or add a method that connect the signal to given callable.

KoBeWi commented on editor/window_wrapper.h:
```diff
@@ -0,0 +1,99 @@
+	GDCLASS(WindowWrapper, MarginContainer);
+
+	Control *wrapped_control = nullptr;
+	MarginContainer *margins = nullptr;
```
WindowWrapper is already a MarginContainer, so what is this for?

trollodel replied:
WindowWrapper is a MarginContainer just because I need a container and MarginContainer is the simplest.
The real window margins are applied using `margins`, which has `window` as parent.

KoBeWi replied:
But you can just set wrapper's margins instead of adding another control.

trollodel replied:
Let me explain the node layout of WindowWrapper.

```
WindowWrapper:
    wrapped_control (when docked)
    window
        window_background
        margins
            wrapped_control (when undocked)
```

As you can see, WindowWrapper and margins are in different windows, and used in different situations.

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,397 @@
+		Rect2 control_rect = _get_default_window_rect();
+
+		// Move the control to the window.
+		remove_child(wrapped_control);
+		parent->add_child(wrapped_control);
```
```suggestion
		wrapped_control->reparent(parent, false);
```
You can use the new `reparent()` method (applies to everywhere you reparent).

KoBeWi commented:
What is this popup for?
![image](https://user-images.githubusercontent.com/2223172/234414962-cb1a7bc7-a53e-422a-a2b1-efd75b15f45e.png)
Pressing the buttons does nothing. Why is the current screens a star? You could change font color instead. Also the popup too hidden; needs to be mentioned in a tooltip probably.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -4954,6 +4959,37 @@ void EditorNode::_save_docks_to_config(Ref<ConfigFile> p_layout, const String &p
+	Dictionary floating_docks_dump;
+
+	for (Control *floating_dock : floating_docks) {
+		WindowWrapper *wrapper = Object::cast_to<WindowWrapper>(floating_dock);
```
`floating_docks` can be changed to `Vector<WindowWrapper>`, so you wouldn't need to cast it.

trollodel replied:
TBH, the whole dock code stuff should be rewritten to always wrap docks in WindowWrapper, like what I did in the Script and Shader editors.

trollodel commented:
> Pressing the buttons does nothing. Why is the current screens a star? You could change font color instead. Also the popup too hidden; needs to be mentioned in a tooltip probably.

Yeah, I was never convinced of that popup. 
Didn't know which was the convention of "currently selected" so I put a star :smile: .
Discoverability is not existing, as you noted. I would put a little arrow pointing down in the right of the button, but was too difficult to do right, so I ignored that. But a tooltip makes sense.

>  Ok I figured out the intended functionality, but it doesn't work xd

Needs to test this again with 2 monitors. But with a single monitor seems to work.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -5100,7 +5136,57 @@ void EditorNode::_dock_tab_changed(int p_tab) {
+
 void EditorNode::_load_docks_from_config(Ref<ConfigFile> p_layout, const String &p_section) {
+	Dictionary floating_docks_dump;
+	if (p_layout->has_section_key(p_section, "dock_floating")) {
+		floating_docks_dump = p_layout->get_value(p_section, "dock_floating");
+	}
```
Can be simplified to
```suggestion
	floating_docks_dump = p_layout->get_value(p_section, "dock_floating", Dictionary());
```

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,397 @@
+
+	set_process_shortcut_input(true);
+
+	EDITOR_DEF_RST("interface/multi_window/enable", true);
+	EDITOR_DEF("interface/multi_window/restore_windows_on_load", true);
+	EDITOR_DEF("interface/multi_window/maximize_window", false);
```
It's better to move these to `EditorSettings::_load_defaults`, so they can be picked by doctool.

KoBeWi commented:
> I test it again and the feature is completely broken now, and put the window in the top left of the screen. Need to fix this.
The intended feature is to place the window to match the control rect when it's docked (see the video in the description). I didn't think about remembering the last window rect.

It's working for me (Windows). The only problem is this:

https://user-images.githubusercontent.com/2223172/234421759-a7d76bfa-e722-44c2-a114-b40ac0d74a16.mp4

Notice how the window appears at the remembered position and then gets forced to the default one.

Although the "window in top left of the screen" does happen with docks. But only for a moment, they get teleported to the default position, just like the script editor.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -5124,10 +5210,21 @@ void EditorNode::_load_docks_from_config(Ref<ConfigFile> p_layout, const String
+				continue;
+			}
+
+			WindowWrapper *wrapper = Object::cast_to<WindowWrapper>(node);
+			if (wrapper) {
+				_dock_floating_close_request(false, wrapper);
+			}
```
I assume this is to unfloat docks when changing editor layout?
It doesn't work.

trollodel replied:
The whole editor layout stuff seems to be broken and only functional for dock position and to save the last editor layout and restore on startup. I didn't dig deeper but seems that this function and the `set_window_layout` methods in plugins aren't called.

KoBeWi replied:
It's not a startup issue though. When you make a dock floating and use Editor -> Editor Layout -> Default, it's supposed to restore all floating docks to their original position. It doesn't happen.

trollodel replied:
Turns out that I'm dumb and that code is wrong. But there is still something wrong in `set_window_layout`. I need to investigate more.

KoBeWi commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -40,7 +40,9 @@
 #include "core/version.h"
 #include "editor/debugger/editor_debugger_node.h"
 #include "editor/debugger/script_editor_debugger.h"
+#include "editor/editor_command_palette.h"
 #include "editor/editor_help_search.h"
+#include "editor/editor_interface.h"
```
Where does this come from?

KoBeWi commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -1593,21 +1596,6 @@ void ScriptEditor::_notification(int p_what) {
 	switch (p_what) {
 		case NOTIFICATION_ENTER_TREE: {
 			EditorNode::get_singleton()->connect("stop_pressed", callable_mp(this, &ScriptEditor::_editor_stop));
```
Missed one connect.

KoBeWi commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -4097,6 +4115,41 @@ ScriptEditor::~ScriptEditor() {
+void ScriptEditorPlugin::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE: {
+			connect("main_screen_changed",
+					callable_mp(this, &ScriptEditorPlugin::_save_last_editor));
```
```suggestion
			connect("main_screen_changed", callable_mp(this, &ScriptEditorPlugin::_save_last_editor));
```
It's shorter than the lines above, so why break it 🙃

KoBeWi commented:
The editor doesn't launch in single-window mode (`--single-window`).
```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.1.dev.custom_build (e0e93ce094d19d132f34aa5c92e892524a8125c9)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] Node::add_child (C:\godot_source\scene\main\node.cpp:1156)
[1] WindowWrapper::set_margins_enabled (C:\godot_source\editor\window_wrapper.cpp:272)
[2] ScriptEditorPlugin::ScriptEditorPlugin (C:\godot_source\editor\plugins\script_editor_plugin.cpp:4291)
[3] EditorNode::EditorNode (C:\godot_source\editor\editor_node.cpp:7993)
[4] Main::start (C:\godot_source\main\main.cpp:2868)
[5] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:178)
[6] _main (C:\godot_source\platform\windows\godot_windows.cpp:203)
[7] main (C:\godot_source\platform\windows\godot_windows.cpp:217)
[8] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:231)
[9] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[10] <couldn't map PC to fn name>
-- END OF BACKTRACE --
================================================================
```

KoBeWi commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -4160,10 +4216,63 @@ void ScriptEditorPlugin::apply_changes() {
+
+	if (EDITOR_GET("interface/multi_window/restore_windows_on_load") && window_wrapper->is_window_available() && p_layout->has_section_key("ScriptEditor", "window_rect")) {
+		// Restore the detached window.
+		Rect2i window_rect = p_layout->get_value("ScriptEditor", "window_rect", Rect2i());
+		int screen = p_layout->get_value("ScriptEditor", "window_screen", -1);
+		Rect2i screen_rect = p_layout->get_value("ScriptEditor", "window_screen_rect", Rect2i());
```
This code seems to repeat in 3 different places. Maybe it could be a method of WindowWrapper and the 3 settings could be its arguments.

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,397 @@
+
+// WindowWrapper
+
+// Capture all the shortcut event not handled by other nodes.
```
```suggestion
// Capture all the shortcut events not handled by other nodes.
```

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,397 @@
+	_populate_screen_list(screen_list);
+}
+
+void ScreenSelect::_populate_screen_list(Container *p_container) {
```
Why is this a separate method if it's called only once?

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,397 @@
+
+// ScreenSelect
+
+void ScreenSelect::_build_advanced_menu() {
```
This method is wrong. It's constantly adding new nodes, without freeing the old ones. Every time the popup is opened, it has more children.

trollodel replied:
Now it should cleanup the popup on close.

KoBeWi commented:
Ok I finished the first pass of reviewing.
The implementation looks fine overall. The main problems that should be solved are:
- single-window mode crashing at startup
- ScreenSelect popup not working correctly
From less important stuff:
- window layout weirdness (https://github.com/godotengine/godot/pull/62378#issuecomment-1522513413)

Also when script editor is detached, drag and drop doesn't function correctly. But there are some PR that will handle it (#71509, #67531)

trollodel commented:
Ok, I addressed most review comments. I also clean up the code a bit, add docs for the new editor settings and test again the features (including the one described in [comment](https://github.com/godotengine/godot/pull/62378#issuecomment-1522485166)).

<hr>

I need to workaround a crash [here](https://github.com/godotengine/godot/pull/62378/files#diff-98ae49b43f623c9f779ac51456c6778f8a51f823ea7f83f60582e65d49ac2991L554-R558) to make the screen selector works, but I don't really understand why it happens. If that workaround is not acceptable, I can disable the feature until it is fixed.

<details>
    <summary>Full stack trace (without the changes in window.cpp)</summary>

    ```
    [1] /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f740ae42520] (??:0)
    [2] Window::_make_window() (/godot/scene/main/window.cpp:556)
    [3] Window::set_visible(bool) (/godot/scene/main/window.cpp:740)
    [4] WindowWrapper::_set_window_enabled_with_rect(bool, Rect2, bool) (/godot/editor/window_wrapper.cpp:140)
    [5] WindowWrapper::enable_window_on_screen(int, bool) (??:?)
    [6] CallableCustomBind::call(Variant const**, int, Variant&, Callable::CallError&) const (/godot/core/variant/callable_bind.cpp:145)
    [7] Object::emit_signalp(StringName const&, Variant const**, int) (/godot/core/object/object.cpp:1056)
    [8] Error Object::emit_signal<int>(StringName const&, int) (??:?)
    [9] ScreenSelect::_emit_screen_signal(int) (/godot/core/string/string_name.h:183)
    [10] CallableCustomBind::call(Variant const**, int, Variant&, Callable::CallError&) const (/godot/core/variant/callable_bind.cpp:145)
    [11] Object::emit_signalp(StringName const&, Variant const**, int) (/godot/core/object/object.cpp:1056)
    [12] BaseButton::_pressed() (/godot/scene/gui/base_button.cpp:139)
    [13] BaseButton::on_action_event(Ref<InputEvent>) (/godot/core/object/ref_counted.h:101)
    [14] BaseButton::gui_input(Ref<InputEvent> const&) (/godot/core/object/ref_counted.h:220)
    [15] Control::_call_gui_input(Ref<InputEvent> const&) (/godot/scene/gui/control.cpp:1752)
    [16] Viewport::_gui_call_input(Control*, Ref<InputEvent> const&) (/godot/scene/main/node.h:353)
    [17] Viewport::_gui_input_event(Ref<InputEvent>) (??:?)
    [18] Viewport::push_input(Ref<InputEvent> const&, bool) (/godot/core/object/ref_counted.h:220)
    [19] Window::_window_input(Ref<InputEvent> const&) (/godot/scene/main/window.cpp:1388)
    [20] void call_with_variant_args_helper<Window, Ref<InputEvent> const&, 0ul>(Window*, void (Window::*)(Ref<InputEvent> const&), Variant const**, Callable::CallError&, IndexSequence<0ul>) (/godot/core/object/ref_counted.h:220)
    [21] DisplayServerX11::_dispatch_input_event(Ref<InputEvent> const&) (??:?)
    [22] Input::_parse_input_event_impl(Ref<InputEvent> const&, bool) (??:?)
    [23] Input::flush_buffered_events() (/godot/core/templates/list.h:351)
    [24] DisplayServerX11::process_events() (/godot/core/templates/local_vector.h:326)
    [25] OS_LinuxBSD::run() (/godot/platform/linuxbsd/os_linuxbsd.cpp:887)
    [26] /godot/bin/godot.linuxbsd.editor.dev.x86_64.llvm(main+0xcb) [0x55d17e81cd7b] (/godot/platform/linuxbsd/godot_linuxbsd.cpp:75)
    [27] /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f740ae29d90] (??:0)
    [28] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7f740ae29e40] (??:0)
    [29] /godot/bin/godot.linuxbsd.editor.dev.x86_64.llvm(_start+0x25) [0x55d17e81cbe5] (??:?)
    ```
</details>

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,450 @@
+// ScreenSelect
+
+void ScreenSelect::_build_advanced_menu() {
+	// Create the popup
```
```suggestion
	// Create the popup.
```
Same below.

KoBeWi commented:
The crash and weird positioning seem to be fixed, but the issue with screen select popup still remains (pressing the buttons does nothing).

Oh and I just found a crash:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.1.dev.custom_build (8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] Control::get_rect (C:\godot_source\scene\gui\control.cpp:1449)
[1] Control::get_rect (C:\godot_source\scene\gui\control.cpp:1449)
[2] PopupMenu::gui_input (C:\godot_source\scene\gui\popup_menu.cpp:400)
[3] call_with_variant_args_helper<PopupMenu,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:293)
[4] call_with_variant_args<PopupMenu,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:408)
[5] CallableCustomMethodPointer<PopupMenu,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[6] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[7] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1056)
[8] Object::emit_signal<Ref<InputEvent> > (C:\godot_source\core\object\object.h:868)
[9] Window::_window_input (C:\godot_source\scene\main\window.cpp:1386)
[10] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:293)
[11] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:408)
[12] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[13] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[14] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:2297)
[15] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:2262)
[16] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:690)
[17] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:941)
[18] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:1991)
[19] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:1475)
[20] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:181)
[21] _main (C:\godot_source\platform\windows\godot_windows.cpp:203)
[22] main (C:\godot_source\platform\windows\godot_windows.cpp:217)
[23] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:231)
[24] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[25] <couldn't map PC to fn name>
-- END OF BACKTRACE --
```
When clicking a button in the screen popup after you make its wrapper float and unfloat (I think).

Also please change that weird star >_> As I said, it can be a colored number instead (e.g. using accent color).

trollodel commented:
> The crash and weird positioning seem to be fixed, but the issue with screen select popup still remains (pressing the buttons does nothing).

I tested it and works, I don't really understand why it doesn't work for you. Can you make a video please?

> Oh and I just found a crash:
> 
> ```
> CrashHandlerException: Program crashed
> Engine version: Godot Engine v4.1.dev.custom_build (8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b)
> Dumping the backtrace. Please include this when reporting the bug to the project developer.
> [0] Control::get_rect (C:\godot_source\scene\gui\control.cpp:1449)
> [1] Control::get_rect (C:\godot_source\scene\gui\control.cpp:1449)
> [2] PopupMenu::gui_input (C:\godot_source\scene\gui\popup_menu.cpp:400)
> [3] call_with_variant_args_helper<PopupMenu,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:293)
> [4] call_with_variant_args<PopupMenu,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:408)
> [5] CallableCustomMethodPointer<PopupMenu,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
> [6] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
> [7] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1056)
> [8] Object::emit_signal<Ref<InputEvent> > (C:\godot_source\core\object\object.h:868)
> [9] Window::_window_input (C:\godot_source\scene\main\window.cpp:1386)
> [10] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:293)
> [11] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:408)
> [12] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
> [13] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
> [14] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:2297)
> [15] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:2262)
> [16] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:690)
> [17] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:941)
> [18] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:1991)
> [19] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:1475)
> [20] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:181)
> [21] _main (C:\godot_source\platform\windows\godot_windows.cpp:203)
> [22] main (C:\godot_source\platform\windows\godot_windows.cpp:217)
> [23] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:231)
> [24] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
> [25] <couldn't map PC to fn name>
> -- END OF BACKTRACE --
> ```
> 
> When clicking a button in the screen popup after you make its wrapper float and unfloat (I think).

I reproduced it and fixed.

> Also please change that weird star >_> As I said, it can be a colored number instead (e.g. using accent color).

Oh, I forgot about that, now it uses the accent color as you suggested.

KoBeWi commented:
> I tested it and works, I don't really understand why it doesn't work for you. Can you make a video please?

Well what I said

https://user-images.githubusercontent.com/2223172/236200125-759217b7-e93c-43d1-ba4e-763a770ebc2e.mp4

The buttons do nothing. I also noticed they are supposed to close the popup, but it only happens after second click.
This is on Windows.

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,442 @@
+	vb->set_alignment(BoxContainer::ALIGNMENT_CENTER);
+	popup_root->add_child(vb);
+
+	Label *description = memnew(Label(TTR("Select screen")));
```
```suggestion
	Label *description = memnew(Label(TTR("Select Screen")));
```
We seem to be using this casing everywhere.

KoBeWi commented on doc/classes/EditorSettings.xml:
```diff
@@ -577,6 +577,17 @@
 		</member>
+		<member name="interface/multi_window/enable" type="bool" setter="" getter="">
+			If [code]true[/code], the multi window support in editor is enabled. The following panels can become dedicated windows (made floating): Docks, Script editor, and Shader editor.
+			When [code]false[/code], modal windows and tooltips are separate windows and displayable outside the editor window.
```
This line is misleading. Windows and tooltips are separate regardless of this setting.

KoBeWi commented:
I noticed that floating docks are no longer remembered after restart. I was working before and it works with script editor.

KoBeWi commented on doc/classes/EditorSettings.xml:
```diff
@@ -577,6 +577,17 @@
 		</member>
+		<member name="interface/multi_window/enable" type="bool" setter="" getter="">
+			If [code]true[/code], the multi window support in editor is enabled. The following panels can become dedicated windows (made floating): Docks, Script editor, and Shader editor.
+			[b]NOTE:[/b] When [member "interface/editor/single_window_mode"] is true, the multi window support is always disabled.
```
```suggestion
			[b]Note:[/b] When [member "interface/editor/single_window_mode"] is true, the multi window support is always disabled.
```

KoBeWi commented:
Found another crash:
```
_realloc_base(void * block, unsigned __int64 size) Line 46 (c:\Users\Tomek\AppData\Local\Programs\Microsoft VS Code\minkernel\crts\ucrt\src\appcrt\heap\realloc_base.cpp:46)
Memory::realloc_static(void * p_memory, unsigned __int64 p_bytes, bool p_pad_align) Line 129 (c:\godot_source\core\os\memory.cpp:129)
CowData<ObjectID>::resize<0>(int p_size) Line 296 (c:\godot_source\core\templates\cowdata.h:296)
Vector<ObjectID>::resize(int p_size) Line 91 (c:\godot_source\core\templates\vector.h:91)
Vector<ObjectID>::push_back(ObjectID p_elem) Line 313 (c:\godot_source\core\templates\vector.h:313)
Vector<ObjectID>::append(const ObjectID & p_elem) Line 70 (c:\godot_source\core\templates\vector.h:70)
SceneTree::_call_input_pause(const StringName & p_group, SceneTree::CallInputType p_call_type, const Ref<InputEvent> & p_input, Viewport * p_viewport) Line 931 (c:\godot_source\scene\main\scene_tree.cpp:931)
Viewport::push_unhandled_input(const Ref<InputEvent> & p_event, bool p_local_coords) Line 2897 (c:\godot_source\scene\main\viewport.cpp:2897)
ShortcutBin::shortcut_input(const Ref<InputEvent> & p_event) Line 70 (c:\godot_source\editor\window_wrapper.cpp:70)
Node::_call_shortcut_input(const Ref<InputEvent> & p_event) Line 2811 (c:\godot_source\scene\main\node.cpp:2811)
SceneTree::_call_input_pause(const StringName & p_group, SceneTree::CallInputType p_call_type, const Ref<InputEvent> & p_input, Viewport * p_viewport) Line 938 (c:\godot_source\scene\main\scene_tree.cpp:938)
Viewport::push_unhandled_input(const Ref<InputEvent> & p_event, bool p_local_coords) Line 2897 (c:\godot_source\scene\main\viewport.cpp:2897)
ShortcutBin::shortcut_input(const Ref<InputEvent> & p_event) Line 70 (c:\godot_source\editor\window_wrapper.cpp:70)
Node::_call_shortcut_input(const Ref<InputEvent> & p_event) Line 2811 (c:\godot_source\scene\main\node.cpp:2811)
SceneTree::_call_input_pause(const StringName & p_group, SceneTree::CallInputType p_call_type, const Ref<InputEvent> & p_input, Viewport * p_viewport) Line 938 (c:\godot_source\scene\main\scene_tree.cpp:938)
```
Seems to be happening when I press Shift in the floating script editor. There is no usual stack trace, I had to use debugger.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -4940,24 +4984,33 @@ void EditorNode::_load_docks_from_config(Ref<ConfigFile> p_layout, const String
 
-			if (dock_slot[atidx]->get_tab_count() == 0) {
-				dock_slot[atidx]->hide();
+			WindowWrapper *wrapper = Object::cast_to<WindowWrapper>(node);
```
This breaks loading the floating docks. `node` is not a WindowWrapper, it's the dock itself.

KoBeWi commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -4174,15 +4252,22 @@ void ScriptEditorPlugin::edited_scene_changed() {
 	EDITOR_GET("text_editor/behavior/files/auto_reload_scripts_on_external_change");
 	ScriptServer::set_reload_scripts_on_save(EDITOR_DEF("text_editor/behavior/files/auto_reload_and_parse_scripts_on_save", true));
 	EDITOR_DEF("text_editor/behavior/files/open_dominant_script_on_scene_change", true);
+	window_wrapper->connect("window_visibility_changed", callable_mp(this, &ScriptEditorPlugin::_window_visibility_changed));
```
Shouldn't this be a few lines higher? (after `wrapper->hide()`)
Weird place to put this line.

trollodel replied:
It's git that likes to move lines :)

trollodel commented:
> Found another crash:
> 
> ```
> _realloc_base(void * block, unsigned __int64 size) Line 46 (c:\Users\Tomek\AppData\Local\Programs\Microsoft VS Code\minkernel\crts\ucrt\src\appcrt\heap\realloc_base.cpp:46)
> Memory::realloc_static(void * p_memory, unsigned __int64 p_bytes, bool p_pad_align) Line 129 (c:\godot_source\core\os\memory.cpp:129)
> CowData<ObjectID>::resize<0>(int p_size) Line 296 (c:\godot_source\core\templates\cowdata.h:296)
> Vector<ObjectID>::resize(int p_size) Line 91 (c:\godot_source\core\templates\vector.h:91)
> Vector<ObjectID>::push_back(ObjectID p_elem) Line 313 (c:\godot_source\core\templates\vector.h:313)
> Vector<ObjectID>::append(const ObjectID & p_elem) Line 70 (c:\godot_source\core\templates\vector.h:70)
> SceneTree::_call_input_pause(const StringName & p_group, SceneTree::CallInputType p_call_type, const Ref<InputEvent> & p_input, Viewport * p_viewport) Line 931 (c:\godot_source\scene\main\scene_tree.cpp:931)
> Viewport::push_unhandled_input(const Ref<InputEvent> & p_event, bool p_local_coords) Line 2897 (c:\godot_source\scene\main\viewport.cpp:2897)
> ShortcutBin::shortcut_input(const Ref<InputEvent> & p_event) Line 70 (c:\godot_source\editor\window_wrapper.cpp:70)
> Node::_call_shortcut_input(const Ref<InputEvent> & p_event) Line 2811 (c:\godot_source\scene\main\node.cpp:2811)
> SceneTree::_call_input_pause(const StringName & p_group, SceneTree::CallInputType p_call_type, const Ref<InputEvent> & p_input, Viewport * p_viewport) Line 938 (c:\godot_source\scene\main\scene_tree.cpp:938)
> Viewport::push_unhandled_input(const Ref<InputEvent> & p_event, bool p_local_coords) Line 2897 (c:\godot_source\scene\main\viewport.cpp:2897)
> ShortcutBin::shortcut_input(const Ref<InputEvent> & p_event) Line 70 (c:\godot_source\editor\window_wrapper.cpp:70)
> Node::_call_shortcut_input(const Ref<InputEvent> & p_event) Line 2811 (c:\godot_source\scene\main\node.cpp:2811)
> SceneTree::_call_input_pause(const StringName & p_group, SceneTree::CallInputType p_call_type, const Ref<InputEvent> & p_input, Viewport * p_viewport) Line 938 (c:\godot_source\scene\main\scene_tree.cpp:938)
> ```
> 
> Seems to be happening when I press Shift in the floating script editor. There is no usual stack trace, I had to use debugger.
> 
> EDIT: Seems like docks also have this problem.

Fixed. It was an infinite recursion, and it was happening with any shortcut.

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,429 @@
+
+		button->set_text(itos(i));
+		button->set_text_alignment(HORIZONTAL_ALIGNMENT_CENTER);
+		button->set_tooltip_text(vformat(TTR("Make this panel floating in the screen %d"), i));
```
```suggestion
		button->set_tooltip_text(vformat(TTR("Make this panel floating in the screen %d."), i));
```
or, but not sure if it's correct:
```suggestion
		button->set_tooltip_text(vformat(TTR("Make this panel floating on screen %d."), i));
```

KoBeWi commented:
I found out why the screen popup doesn't do anything, but I don't know what causes it .-.
The buttons do not emit the `pressed` signal. It's as if they don't receive clicks. It *might* be because you're abusing OptionButton's popup for that instead of a custom one, but not sure why would it make difference, so maybe it's not that. But the buttons are unclickable.
Which is weird, because you said it works for you.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -4940,24 +4984,31 @@ void EditorNode::_load_docks_from_config(Ref<ConfigFile> p_layout, const String
 				break;
 			}
+
 			if (atidx == -1) { // Well, it's not anywhere.
```
The problem of dock not getting unfloated on layout change comes from here. This code only looks for docks in dock slots, but floating docks are elsewhere. You need to add a search in `floating_docks` too.

trollodel replied:
Now I get what you have tested and why it doesn't work. I didn't think about restoring layout with docks currently floating.
I'll look to it.

KoBeWi commented:
I did some testing regarding the screen popup issue. The buttons don't receive pressed mouse events. Seems like a problem exclusive to PopupMenu. I guess using MenuButton this way is wrong. So instead ScreenSelect should be a regular Button with a custom PopupPanel.

btw the F1 shortcut to open help does not work for some reason.

trollodel commented:
> I found out why the screen popup doesn't do anything, but I don't know what causes it .-. The buttons do not emit the `pressed` signal. It's as if they don't receive clicks. It _might_ be because you're abusing OptionButton's popup for that instead of a custom one, but not sure why would it make difference, so maybe it's not that. But the buttons are unclickable. Which is weird, because you said it works for you.

> I did some testing regarding the screen popup issue. The buttons don't receive pressed mouse events. Seems like a problem exclusive to PopupMenu. I guess using MenuButton this way is wrong. So instead ScreenSelect should be a regular Button with a custom PopupPanel.

Ok now it uses regular Button and Popup, I should work now for you.

KoBeWi commented:
The popup is now functional. The only problem is that it doesn't close once clicked 🙃

https://user-images.githubusercontent.com/2223172/236520159-ae08ffe4-2dcf-47d0-9738-3885600cd704.mp4

The F1 shortcut is still broken. It needs to be fixed.

Also there are some conflicts, so the PR needs rebase.

trollodel commented:
> The popup is now functional. The only problem is that it doesn't close once clicked upside_down_face

I can't reproduce it either on Linux. Just in case, I forcefully hide the popup when the button is pressed.

> The F1 shortcut is still broken. It needs to be fixed.

After the rebase, the F1 shortcuts now work when docked and undocked.

> Also there are some conflicts, so the PR needs rebase.

Done.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -4940,24 +4986,45 @@ void EditorNode::_load_docks_from_config(Ref<ConfigFile> p_layout, const String
+			if (atidx == -1) {
+				// Try floating docks.
+				for (WindowWrapper *wrapper : floating_docks) {
+					if (wrapper->get_meta("dock_name") == name) {
```
This doesn't work, because `wrapper->get_meta("dock_name")` is StringName and `name` is String.

trollodel replied:
This and other fixes should be enough to make the feature finally working.

KoBeWi commented:
I found an edge case. When you make docks floating and restart editor in single mode, the docks will be unusable:
![image](https://github.com/godotengine/godot/assets/2223172/e6ed0433-cc28-4a13-b527-828499c91969)
Not critical though.

KoBeWi reviewed (approved):
All obvious/important issues are fixed.

akien-mga reviewed (approved):
Looks good to me from a cursory code check, and I tested the functionality a bit and it seems to work well.

I think it's time for a merge and get it tested more widely :)

For the record, I'll release 4.1-dev2 tomorrow but I already built it from a commit prior to merging this, so this will be in an official build in 4.1-dev3.

KoBeWi commented on editor/window_wrapper.cpp:
```diff
@@ -0,0 +1,473 @@
+	set_action_mode(ACTION_MODE_BUTTON_PRESS);
+
+	// Create the popup.
+	const real_t popup_height = real_t(get_theme_font_size("font_size")) * 2.0;
```
The warning might come come from here.

trollodel replied:
It was exactly that.

trollodel commented on editor/editor_node.cpp:
```diff
@@ -4468,67 +4469,66 @@ void EditorNode::_copy_warning(const String &p_str) {
+	wrapper->set_wrapped_control(p_dock);
+	wrapper->set_meta("dock_slot", p_slot_index);
+	wrapper->set_meta("dock_index", dock_index);
+	wrapper->set_meta("dock_name", p_dock->get_name());
```
IMO, It's just moving the inefficiency from one place to another. But fine, I did that.

KoBeWi replied:
The other conversion was in a for loop, which would perform this multiple times for the same dock. Here it happens only once.

trollodel commented:
> For the record, I'll release 4.1-dev2 tomorrow but I already built it from a commit prior to merging this, so this will be in an official build in 4.1-dev3.

So sad, users won't have this absolutely desired feature tomorrow, they would wait for other 2 weeks (sarcasm).

> As a side note, while testing this PR I saved a project with 2 floating docks, which were properly remembered across sessions. Then I reopened that same project in a Godot build from the `master` branch without this PR, and the floating docks were still floating and functional. Pretty cool feature that works retroactively :P

I tested myself and I don't really understand why it works. I guess the docks were placed in the default slot as a failsafe.

AThousandShips commented:
CI should be fixed by #76885

akien-mga commented:
Congrats on landing this amazing feature! :tada: 
Thanks for your patience with the review and release processes :)
